### PR TITLE
fiber: hide backtraces of idle fibers from fiber.info

### LINKED ISSUE #4235: Don't show idle fibers in fiber.info
Output of `fiber.info` from an instance that has been running for a while:
```Yaml
    name: main
  3965366:
    csw: 3502
    backtrace:
    - C: '#0  0x40efdc in fiber_cxx_invoke(int (*)(__va_list_tag*), __va_list_tag*)+12'
    - C: '#1  0x4f7e90 in fiber_loop+48'
    - C: '#2  0x6323bf in coro_init+63'
    fid: 3965366
    memory:
      total: 545104
      used: 23926
    name: main
  3965361:
    csw: 24319
    backtrace:
    - C: '#0  0x40efdc in fiber_cxx_invoke(int (*)(__va_list_tag*), __va_list_tag*)+12'
    - C: '#1  0x4f7e90 in fiber_loop+48'
    - C: '#2  0x6323bf in coro_init+63'
    fid: 3965361
    memory:
      total: 536912
      used: 15917
    name: main
  3974361:
    csw: 53165
    backtrace:
    - C: '#0  0x40efdc in fiber_cxx_invoke(int (*)(__va_list_tag*), __va_list_tag*)+12'
    - C: '#1  0x4f7e90 in fiber_loop+48'
    - C: '#2  0x6323bf in coro_init+63'
    fid: 3974361
    memory:
      total: 520528
      used: 0
    name: main
  3965362:
    csw: 4138
    backtrace:
    - C: '#0  0x40efdc in fiber_cxx_invoke(int (*)(__va_list_tag*), __va_list_tag*)+12'
    - C: '#1  0x4f7e90 in fiber_loop+48'
    - C: '#2  0x6323bf in coro_init+63'
    fid: 3965362
    memory:
      total: 541008
      used: 20066
    name: main
  3965372:
    csw: 616
    backtrace:
    - C: '#0  0x40efdc in fiber_cxx_invoke(int (*)(__va_list_tag*), __va_list_tag*)+12'
    - C: '#1  0x4f7e90 in fiber_loop+48'
    - C: '#2  0x6323bf in coro_init+63'
    fid: 3965372
    memory:
      total: 536912
      used: 15936
```
We need to filter out idle fiber stacks, because they only clutter the output.

### PR DESCRIPTION
**Pull request by Egor2001**
`FIBER_IS_IDLE` flag is added to track fibers from fiber pool.
These fibers are still present in the `fiber.info` but without their stacks. 

Closes #4235

### CODE REVIEW

**Comment by Egor2001** [comment]
Basically, one of the 3 possible approaches of handling idle fibers in the `fiber.info` may be implemented:
1. Not to show idle fibers in info at all unless the corresponding flag is set to true.
2. Unconditionally hide only idle fiber stacks as they are the most cluttering in the `fiber.info`.
3. To combine 2 approaches above.

Currently this PR uses the 2nd approach.
Can this approach be taken, @Mons?

**Comment by Gerold103** [comment]
I see the last commit undoes the core idea of the first commit. Please, decide on the final solution before a detailed review would be started.

**Comment by Mons** [comment]
I consider, that both ideas are good
It is worth hiding idle fiber stacks.
It is also worth hiding idle fibers under some flag of `fiber.info`

**Comment by Egor2001** [comment]
> I see the last commit undoes the core idea of the first commit. Please, decide on the final solution before a detailed review would be started.

The final solution is to hide only stacks of idle fibers, as @Mons accepted the current idea.

**Comment by Gerold103** [comment]
Is it ready for a review? I still see the last commit undoes part of the first commit.

**Comment by Egor2001** [comment]
> I still see the last commit undoes a part of the first commit.

Fixed

File: test/app/gh-4235-dont-show-idle-in-fiber-info.result [hunk 1]
```diff
@@ -0,0 +1,104 @@
+-- test-run result file version 2
```
Comment (Gerold103) [inline test/app/gh-4235-dont-show-idle-in-fiber-info.result:1]: The test failed one time on my machine:
```
[009] app/gh-4235-dont-show-idle-in-fiber-info.test.>                 [ fail ]
[009] 
[009] Test failed! Result content mismatch:
[009] --- app/gh-4235-dont-show-idle-in-fiber-info.result	Tue Jun  8 21:26:38 2021
[009] +++ var/rejects/app/gh-4235-dont-show-idle-in-fiber-info.reject	Tue Jun  8 21:59:27 2021
[009] @@ -89,11 +89,11 @@
[009]  
[009]  new_idle > ndelta
[009]   | ---
[009] - | - true
[009] + | - false
[009]   | ...
[009]  new_all - new_idle < old_all - old_idle + ndelta
[009]   | ---
[009] - | - true
[009] + | - false
[009]   | ...
[009]  
[009]  conn:close()
```
Couldn't reproduce it again. But still, that makes the test flaky. You might need to design a more robust way to test the patch. Not depending on probabilities, magic constants, timeouts, etc.

File: src/lib/core/fiber_pool.c [hunk 2]
```diff
@@ -105,6 +109,7 @@ fiber_pool_idle_cb(ev_loop *loop, struct ev_timer *watcher, int events)
 		 * scheduled lately.
 		 */
 		f = rlist_shift_tail_entry(&pool->idle, struct fiber, state);
+		f->flags &= ~FIBER_IS_IDLE;
```
[the remark this answers is not in the document]
Comment (Egor2001) [reply src/lib/core/fiber_pool.c]: Fixed.

File: src/lib/core/fiber_pool.c [hunk 3]
```diff
@@ -125,6 +130,7 @@ fiber_pool_cb(ev_loop *loop, struct ev_watcher *watcher, int events)
 		struct fiber *f;
 		if (! rlist_empty(&pool->idle)) {
 			f = rlist_shift_entry(&pool->idle, struct fiber, state);
+			f->flags &= ~FIBER_IS_IDLE;
```
[the remark this answers is not in the document]
Comment (Egor2001) [reply src/lib/core/fiber_pool.c]: Fixed.

File: src/lib/core/fiber_pool.c [hunk 4]
```diff
@@ -185,8 +191,10 @@ fiber_pool_destroy(struct fiber_pool *pool)
 	 */
 	pool->idle_timeout = 0;
 	struct fiber *idle_fiber;
-	rlist_foreach_entry(idle_fiber, &pool->idle, state)
+	rlist_foreach_entry(idle_fiber, &pool->idle, state) {
+		idle_fiber->flags &= ~FIBER_IS_IDLE;
```
[the remark this answers is not in the document]
Comment (Egor2001) [reply src/lib/core/fiber_pool.c]: Fixed.

File: src/lua/fiber.c [hunk 5]
```diff
@@ -319,7 +319,8 @@ lbox_fiber_statof(struct fiber *f, void *cb_ctx, bool backtrace)
 	struct lua_State *L = cb_ctx;
 	luaL_pushuint64(L, f->fid);
 	lua_newtable(L);
-	lbox_fiber_statof_map(f, cb_ctx, backtrace);
+	lbox_fiber_statof_map(f, cb_ctx, 
+			      backtrace && ! (f->flags & FIBER_IS_IDLE));
```
[the remark this answers is not in the document]
Comment (Egor2001) [reply src/lua/fiber.c]: Thanks, fixed.

File: test/app/gh-4235-dont-show-idle-in-fiber-info.result [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+done = 0
+ | ---
+ | ...
+test_run:cmd('setopt delimiter ";"')
+ | ---
+ | - true
+ | ...
+function yielder()
+    for i = 1,nyields do
```
[the remark this answers is not in the document]
Comment (Egor2001) [reply test/app/gh-4235-dont-show-idle-in-fiber-info.result]: Fixed (test is rewritten).

File: test/app/gh-4235-dont-show-idle-in-fiber-info.result [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        fiber.sleep(yield_sleepval)
+    end done = done + 1
```
[the remark this answers is not in the document]
Comment (Egor2001) [reply test/app/gh-4235-dont-show-idle-in-fiber-info.result]: Fixed (test is rewritten).

File: test/app/gh-4235-dont-show-idle-in-fiber-info.result [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+ | ---
+ | ...
+niters = 100
+ | ---
+ | ...
+yield_sleepval = 0.001
+ | ---
+ | ...
+
+pause_sleepval = 0.2 + 10 * nyields * yield_sleepval
```
[the remark this answers is not in the document]
Comment (Egor2001) [reply test/app/gh-4235-dont-show-idle-in-fiber-info.result]: Fixed. Test is redesigned not to rely on magic constants.

File: test/app/gh-4235-dont-show-idle-in-fiber-info.result [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+ | ---
+ | ...
+
+ -- max increase in fiber.info entries count 
+ | ---
+ | ...
+ -- after the test execution
+ | ---
+ | ...
+ndelta = 5
```
[the remark this answers is not in the document]
Comment (Egor2001) [reply test/app/gh-4235-dont-show-idle-in-fiber-info.result]: Fixed.

File: test/app/gh-4235-dont-show-idle-in-fiber-info.result (hunk 1 above)
Comment (Egor2001) [reply test/app/gh-4235-dont-show-idle-in-fiber-info.result:1]: Thanks for your reply, test is redesigned to be more deterministic.

File: src/lib/core/fiber_pool.c [hunk 10]
```diff
@@ -58,6 +58,7 @@ fiber_pool_f(va_list ap)
 			f->caller = rlist_shift_entry(&pool->idle,
 						      struct fiber,
 						      state);
+			f->caller->flags &= ~FIBER_IS_IDLE;
```
[the remark this answers is not in the document]
Comment (Egor2001) [reply src/lib/core/fiber_pool.c]: Thanks a lot for your test sample!
I really overdid it with dropping the `FIBER_IS_IDLE` flag.
Fixed.

File: test/app/gh-4235-dont-show-idle-in-fiber-info.result [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+ | ---
+ | ...
+netbox = require('net.box')
+ | ---
+ | ...
+fiber = require('fiber')
+ | ---
+ | ...
+
+ -- max increase in fiber.info entries count
```
[the remark this answers is not in the document]
Comment (Egor2001) [reply test/app/gh-4235-dont-show-idle-in-fiber-info.result]: Fixed.

File: src/lua/fiber.c [hunk 12]
```diff
@@ -319,7 +319,8 @@ lbox_fiber_statof(struct fiber *f, void *cb_ctx, bool backtrace)
 	struct lua_State *L = cb_ctx;
 	luaL_pushuint64(L, f->fid);
 	lua_newtable(L);
-	lbox_fiber_statof_map(f, cb_ctx, backtrace);
+	lbox_fiber_statof_map(f, cb_ctx,
```
[the remark this answers is not in the document]
Comment (Egor2001) [reply src/lua/fiber.c]: Fixed.

File: test/app/gh-4235-dont-show-idle-in-fiber-info.result [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+ | ---
+ | ...
+fiber = require('fiber')
+ | ---
+ | ...
+
+ -- max increase in fiber.info entries count 
+ | ---
+ | ...
+ -- after the test execution
```
[the remark this answers is not in the document]
Comment (Egor2001) [reply test/app/gh-4235-dont-show-idle-in-fiber-info.result]: Thanks, fixed.

**Review by Gerold103 — CHANGES_REQUESTED** [review]
Thanks for the fixes! All looks good. Except that neither of your commits closes the issue. One says "Needed for", the other says "Follows up", but I don't see "Closes". The last commit should have it.

Also I don't see a new changelog file. You need to add one to `changelogs/unreleased`.

**Comment by Egor2001** [comment]
Add changelog and fix commit message.

File: test/app/gh-4235-dont-show-idle-in-fiber-info.test.lua [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+end;
+test_run:cmd('setopt delimiter ""');
+
+box.schema.user.grant('guest', 'execute', 'universe')
+conn = netbox.connect(box.cfg.listen)
+
+prev_idle = count_idle()
+
+n = 100
+for i = 1, n do conn:call('yielder', nil, {is_async = true}) end
```
Comment (Korablev77) [inline test/app/gh-4235-dont-show-idle-in-fiber-info.test.lua:34]: Could you please add comment explaining what's going on here. E.g. why do you need netbox? I suspect to generate fibers, but why don't you use fiber.create? Test doesn't seem obvious from the first sight...

File: test/app/gh-4235-dont-show-idle-in-fiber-info.test.lua [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+while cnt < n do fiber.yield() end
+done = true
+while cnt > 0 do fiber.yield() end
+
+-- Most of the fibers created by `conn:call` must be in the IDLE state
+count_idle() >= n - prev_idle
```
Comment (Korablev77) [inline test/app/gh-4235-dont-show-idle-in-fiber-info.test.lua:41]: Nit: use assert to check condition
Nit: this check in the form of count_idle() + prev_idle >= n looks more reasonable (skip this if it doesn't look to you so)

**Comment by Korablev77** [comment]
Two comments above, LGTM otherwise.

**Review by Gerold103 — APPROVED** [review]

File: test/app/gh-4235-dont-show-idle-in-fiber-info.test.lua (hunk 14 above)
Comment (Egor2001) [reply test/app/gh-4235-dont-show-idle-in-fiber-info.test.lua:36]: Added explanatory comment

File: test/app/gh-4235-dont-show-idle-in-fiber-info.test.lua (hunk 15 above)
Comment (Egor2001) [reply test/app/gh-4235-dont-show-idle-in-fiber-info.test.lua]: Thanks, fixed

**Review by Korablev77 — APPROVED** [review]
LGTM

**Pull request merged**